File: Src/ECScape.Core/Systems/RenderSystem.cs
using ECScape.Core.Components;
using ECScape.Core.Engine;
using ECScape.Core.Entities;

namespace ECScape.Core.Systems;

public sealed class RenderSystem : ISystem
{
    private readonly Entry[,] frontBuffer;
    private readonly Entry[,] backBuffer;

    public RenderSystem()
    {
        var width = UiInterface.TotalWidth;
        var height = UiInterface.TotalHeight;

        frontBuffer = new Entry[width, height];
        backBuffer = new Entry[width, height];
        for (var x = 0; x < width; x++)
        {
            for (var y = 0; y < height; y++)
            {
                frontBuffer[x, y] = new(' ');
                backBuffer[x, y] = new(' ');
            }
        }
    }

    public void Update(World world, float deltaTime)
    {
        for (var x = 0; x < backBuffer.GetLength(0); x++)
        {
            for (var y = 0; y < backBuffer.GetLength(1); y++)
            {
                backBuffer[x, y] = new(' ');
            }
        }

        world.Entities
            .Where(x => x.HasComponent<Position>() && x.HasComponent<Drawable>() && x.HasComponent<Exists>()).ToList()
            .ForEach(Draw);

        DrawInterface(world.GetEntityWith(typeof(Statistics)));

        Render();
    }

    private void Draw(Entity entity)
    {
        var drawable = entity.GetRequiredComponent<Drawable>();
        var position = entity.GetRequiredComponent<Position>();
        var size = entity.GetComponent<Size>() ?? new Size(1, 1);

        for (var x = 0; x < size.Width; x++)
        {
            for (var y = 0; y < size.Height; y++)
            {
                var newPos = new Position(position.Left + x, position.Top + y);
                if (newPos.Left < 0 || newPos.Left >= backBuffer.GetLength(0) ||
                    newPos.Top < 0 || newPos.Top >= backBuffer.GetLength(1))
                {
                    continue;
                }

                backBuffer[(int)newPos.Left, (int)newPos.Top] = new(drawable.Symbol, drawable.Color);
            }
        }
    }

    private void DrawInterface(Entity? playerEntity)
    {
        var gameTitle = $"  ECScape  ";
        var titleIndex = 0;
        if (playerEntity != null)
        {
            var score = playerEntity?.GetComponent<Statistics>()?.Score ?? 0;
            var hearts = new string('♥', playerEntity?.GetRequiredComponent<Health>().Hearts ?? 0);
            if (string.IsNullOrWhiteSpace(hearts)) { hearts = "☠"; }
            gameTitle = $"  ECScape  ▧▧▧  Health: {hearts}  ▧▧▧  Score: {score}  ";
        }
        for (var i = 0; i < UiInterface.TotalWidth; i++)
        {
            for (var j = UiInterface.InterfaceStart; j < UiInterface.InterfaceEnd; j++)
            {
                if (i > 3 && j == 1 && titleIndex < gameTitle.Length)
                {
                    var c = gameTitle[titleIndex++];
                    backBuffer[i, j] = new(c, ConsoleColor.Cyan);
                }
                else
                {
                    backBuffer[i, j] = new('▧', ConsoleColor.Gray);
                }
            }
        }
    }

    private void Render()
    {
        for (var x = 0; x < frontBuffer.GetLength(0); x++)
        {
            for (var y = 0; y < frontBuffer.GetLength(1); y++)
            {
                if (frontBuffer[x, y] != backBuffer[x, y])
                {
                    Writer.Write(backBuffer[x, y].Symbol, backBuffer[x, y].Color, new Position(x, y));
                    frontBuffer[x, y] = backBuffer[x, y];
                }
            }
        }
    }

    private readonly record struct Entry(char Symbol, ConsoleColor Color)
    {
        public Entry(char symbol) : this(symbol, ConsoleColor.White) { }
    }
}

File: Web/src/systems/RenderSystem.js
import ISystem from './ISystem.js';
import UiInterface from '../core/uiInterface.js';
import { Position, Size, Drawable, Exists, Statistics, Health } from '../components/Components.js';

class RenderSystem extends ISystem {
    constructor(canvas) {
        super();
        this.canvas = canvas;
        this.ctx = canvas.getContext('2d');
        this.scaleFactor = 1;
        this.resizeCanvas();
        window.addEventListener('resize', () => this.resizeCanvas());
    }

    resizeCanvas() {
        const maxWidth = window.innerWidth;
        const maxHeight = window.innerHeight;

        this.scaleFactor = Math.floor(Math.min(maxWidth / UiInterface.TotalWidth, maxHeight / UiInterface.TotalHeight));

        this.canvas.width = UiInterface.TotalWidth * this.scaleFactor;
        this.canvas.height = UiInterface.TotalHeight * this.scaleFactor;
    }

    update(world, deltaTime) {
        this.ctx.clearRect(0, 0, this.canvas.width, this.canvas.height);
        this.ctx.save();
        this.ctx.scale(this.scaleFactor, this.scaleFactor);

        this.drawTileset(world.Tileset);

        world.getEntitiesWith(Position, Drawable, Exists).forEach(e => this.drawEntity(e));
        this.drawInterface(world.getEntityWith(Statistics));
        this.ctx.restore();
    }

    drawTileset(tileset) {
        if (!tileset) return;

        const tileSize = 32;
        const tilesetImage = document.getElementById('tileset');

        for (let x = 0; x < tileset.length; x++) {
            for (let y = 0; y < tileset[x].length; y++) {
                const tileID = tileset[x][y];
                if (tileID !== 0 && tileID !== 1) continue;

                this.ctx.drawImage(
                    tilesetImage,
                    tileID * tileSize, 0, tileSize, tileSize,
                    x * tileSize, y * tileSize, tileSize, tileSize
                );
            }
        }
    }

    drawEntity(entity) {
        const drawable = entity.getRequiredComponent(Drawable);
        const position = entity.getRequiredComponent(Position);
        const size = entity.getComponent(Size) || new Size();

        this.ctx.fillStyle = drawable.Color;

        if (drawable.Sprite) {
            const spriteWidth = drawable.SpriteWidth;
            const spriteHeight = drawable.SpriteHeight;

            if (size.Width < spriteWidth || size.Height < spriteHeight || !drawable.Tileable) {
                if (position.Direction === 'l') {
                    this.ctx.save();
                    this.ctx.translate(position.Left + size.Width, position.Top);
                    this.ctx.scale(-1, 1);
                    this.ctx.drawImage(
                        drawable.Sprite,
                        0,
                        0,
                        size.Width,
                        size.Height
                    );
                    this.ctx.restore();
                } else {
                    this.ctx.drawImage(
                        drawable.Sprite,
                        position.Left,
                        position.Top,
                        size.Width,
                        size.Height
                    );
                }
            }
            else {
                const tilesX = Math.ceil(size.Width / spriteWidth);
                const tilesY = Math.ceil(size.Height / spriteHeight);

                for (let y = 0; y < tilesY; y++) {
                    const currentY = position.Top + (y * spriteHeight);
                    const currentHeight = y === tilesY - 1
                        ? size.Height - (y * spriteHeight)
                        : spriteHeight;

                    for (let x = 0; x < tilesX; x++) {
                        const currentX = position.Left + (x * spriteWidth);
                        const currentWidth = x === tilesX - 1
                            ? size.Width - (x * spriteWidth)
                            : spriteWidth;

                        this.ctx.drawImage(
                            drawable.Sprite,
                            0, 0,
                            spriteWidth, spriteHeight,
                            currentX, currentY,
                            currentWidth, currentHeight
                        );
                    }
                }
            }
        }
    }

    drawInterface(playerEntity) {
        this.ctx.fillStyle = 'gray';
        this.ctx.fillRect(0, 0, this.canvas.width, UiInterface.InterfaceEnd);

        this.ctx.fillStyle = 'cyan';
        this.ctx.font = 'bold 16px monospace';
        this.ctx.textAlign = 'center';

        let gameTitle = "  ECScape  ";
        if (playerEntity) {
            const score = playerEntity.getComponent(Statistics)?.Score || 0;
            const hearts = '♥'.repeat(playerEntity.getRequiredComponent(Health).Hearts || 0);
            gameTitle = `  ECScape     Health: ${hearts}     Score: ${score}  `;
        }

        this.ctx.fillText(
            gameTitle,
            UiInterface.TotalWidth / 2,
            UiInterface.InterfaceStart + 20
        );
    }
}

export default RenderSystem;